src/DoctrineCheckpointStorage.php

```diff
@@ -32,42 +47,43 @@ public function acquireLock(): SequenceNumber
         } catch (DBALException $exception) {
-            throw new \RuntimeException($exception->getMessage(), 1544207778, $exception);
-        } finally {
             $this->connection->rollBack();
+            throw new \RuntimeException($exception->getMessage(), 1544207778, $exception);
         }
         if (!is_numeric($highestAppliedSequenceNumber)) {
+            $this->connection->rollBack();
             throw new CheckpointException(sprintf('Failed to fetch highest applied sequence number for subscriber "%s". Please run %s::setup()', $this->subscriberId, $this::class), 1652279139);
         }
-        return SequenceNumber::fromInteger((int)$highestAppliedSequenceNumber);
+        $this->lockAcquired = true;
+        $this->lockedSequenceNumber = SequenceNumber::fromInteger((int)$highestAppliedSequenceNumber);
+        return $this->lockedSequenceNumber;
     }
 
     public function updateAndReleaseLock(SequenceNumber $sequenceNumber): void
     {
-        // TODO check for active transaction?
-//        if (!$this->connection->isTransactionActive()) {
-//            throw new CheckpointException(sprintf('Failed to update and commit checkpoint for subscriber "%s" because no transaction is active', $this->subscriberId), 1652279314);
-//        }
-        // Fails if no matching entry exists; which is fine because initializeHighestAppliedSequenceNumber() must be called beforehand.
+        if (!$this->lockAcquired) {
+            throw new CheckpointException(sprintf('Failed to update and commit checkpoint for subscriber "%s" because the lock has not been acquired successfully before', $this->subscriberId), 1660556344);
+        }
+        if (!$this->connection->isTransactionActive()) {
+            throw new CheckpointException(sprintf('Failed to update and commit checkpoint for subscriber "%s" because no transaction is active', $this->subscriberId), 1652279314);
+        }
         try {
-            $this->connection->update(
-                $this->tableName,
-                ['appliedsequencenumber' => $sequenceNumber->value],
-                ['subscriberid' => $this->subscriberId]
-            );
+            if ($this->lockedSequenceNumber === null || !$this->lockedSequenceNumber->equals($sequenceNumber)) {
+                $this->connection->update($this->tableName, ['appliedsequencenumber' => $sequenceNumber->value], ['subscriberid' => $this->subscriberId]);
+            }
+            $this->connection->commit();
         } catch (DBALException $exception) {
+            $this->connection->rollBack();
             throw new CheckpointException(sprintf('Failed to update and commit highest applied sequence number for subscriber "%s". Please run %s::setup()', $this->subscriberId, $this::class), 1652279375, $exception);
+        } finally {
+            $this->lockAcquired = false;
```

Comment:
LockedSequwnceNunber auch zurücksetzen?

Reply:
Yes, good point.. It has no actual effect because it is only set on `acquireLock()` but for good measure.. taking care!